asset-registry/src/lib.rs

```diff
@@ -0,0 +1,235 @@
+use scale_info::TypeInfo;
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, Member},
+	DispatchResult,
+};
+use sp_std::prelude::*;
+use xcm::latest::prelude::*;
+use xcm_builder::TakeRevenue;
+use xcm_executor::{traits::WeightTrader, Assets};
+
+pub trait AssetIdAssigner<AssetId, AssetMetadata> {
+	fn assign_id(asset_metadata: &AssetMetadata, last_id: Option<AssetId>) -> AssetId;
+}
+
+pub struct SequentialId<AssetId: AtLeast32BitUnsigned>(PhantomData<AssetId>);
+impl<AssetId: AtLeast32BitUnsigned, AssetMetadata> AssetIdAssigner<AssetId, AssetMetadata> for SequentialId<AssetId> {
+	fn assign_id(_asset_metadata: &AssetMetadata, last_id: Option<AssetId>) -> AssetId {
+		match last_id {
+			None => AssetId::zero(),
+			Some(x) => x.saturating_add(AssetId::one()),
+		}
+	}
+}
+
+#[frame_support::pallet]
+pub mod module {
+	use super::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type AssetMetadata: Parameter + Member + TypeInfo + Into<MultiLocation>;
+
+		type AssetId: Parameter + Member + TypeInfo;
+
+		type AuthorityOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		type AssignAsset: AssetIdAssigner<Self::AssetId, Self::AssetMetadata>;
+		// /// Weight information for extrinsics in this module.
+		// type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RegisteredAsset {
+			asset_id: T::AssetId,
+			metadata: T::AssetMetadata,
+		},
+	}
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	#[pallet::getter(fn get_metadata)]
+	pub type Metadata<T: Config> = StorageMap<_, Twox64Concat, T::AssetId, T::AssetMetadata, OptionQuery>;
+
+	/// The total issuance of a token type.
+	#[pallet::storage]
+	pub type MultiLocationLookup<T: Config> = StorageMap<_, Twox64Concat, MultiLocation, T::AssetId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn last_asset_id)]
+	pub(crate) type LastAssetId<T: Config> = StorageValue<_, T::AssetId, OptionQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		_phantom: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				_phantom: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {}
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(0)]
+		pub fn register_asset(origin: OriginFor<T>, metadata: T::AssetMetadata) -> DispatchResult {
```

Comment:
I think we can add `asset_id: Option<T::AssetId>` parameter so we can specify the asset id if needed.

Reply:
Agree, parachains like Acala already implements they own asset-registry, If they migrate to orml's asset-registry, seemless migration is helpfull, they could register asset metadata with already exists asset id. In this way, I think `AssetIdAssigner` also needs more trait method for those scene.

Reply:
Done. However, I'm not sure what use case you envision since there is already the `process_asset` trait function that is able to assign an id. Could you clarify?

Reply:
Our asset registry supports not only foreign assets, but also native assets. So that asset_id could be an existing one instead of new id.